flowey/flowey_hvlite/src/pipelines/custom_vmfirmwareigvm_dll.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! See [`CustomVmfirmwareigvmDllCli`]
+
+use crate::pipelines_shared::cfg_common_params::CommonArchCli;
+use anyhow::Context;
+use flowey::node::prelude::ReadVar;
+use flowey::pipeline::prelude::*;
+use std::path::PathBuf;
+
+/// Encapsulate an existing pre-built IGVM file into *unsigned*
+/// `vmfirmwareigvm.dll` resource DLL.
+///
+/// Unlike `build-igvm`, this tool will NOT build OpenHCL from scratch. This
+/// tool streamlines the process of building the in-tree `vmfirmwareigvm_dll`
+/// crate (which requires setting various env vars, installing certain
+/// dependencies, etc...).
+///
+/// NOTE: This tool is primarily intended for use by Microsoft employees, as
+/// open-source deployments of OpenHCL typically load the IGVM file directly
+/// (rather than being encapsulated in a resource-DLL).
+#[derive(clap::Args)]
+pub struct CustomVmfirmwareigvmDllCli {
+    /// Path to IGVM payload to encapsulate in the vmfirmwareigvm resource DLL.
+    pub igvm_payload: PathBuf,
+
+    /// Architecture the DLL should be built for.
+    ///
+    /// Defaults to the current host architecture.
+    #[clap(long)]
+    pub arch: Option<CommonArchCli>,
+}
+
+impl IntoPipeline for CustomVmfirmwareigvmDllCli {
+    fn into_pipeline(self, backend_hint: PipelineBackendHint) -> anyhow::Result<Pipeline> {
+        if !matches!(backend_hint, PipelineBackendHint::Local) {
+            anyhow::bail!("build-igvm is for local use only")
+        }
+
+        // DEVNOTE: it would be nice to figure out what sort of magic is
+        // required for the WSL2 case to work. The tricky part is dealing with
+        // the underlying invocations to `rc.exe` via WSL2.
```

Comment:
maybe @benhillis has ideas? I think if you can somehow invoke the tool not directly, but via a shell so that wsl knows to launch it as a windows exe it would work? i dunno

Reply:
Might be stealing Ben's thunder! This https://man7.org/linux/man-pages/man1/windres.1.html could help. Given that the dll is resource only, the object crate should be able to produce the binary I think

Reply:
"Out of the thin air" is always true for the resource-only DLLs, and I do get why the situation is unsatisfactory.

The object crate is already used for dealing with resource-only dll's (the IGVM "container") [hvlite_pcat_locator/src/resource_dll_parser.rs#L10](https://github.com/microsoft/openvmm/blob/dcce648caa23d6682ce0472f8d4812c366bfb83c/openvmm/hvlite_pcat_locator/src/resource_dll_parser.rs#L10). If you think that an extra push in this PR to resolve the note you left is better left for a later time, opening an issue would be a fair resolution.

Reply:
`object` is great for _reading_ resource DLLs, but it won't get you all the way there with _writing_ resource DLLs.
e.g: there's lots of misc cruft in DLL files that we'd have to manually compose ourselves if we went down this rabbit hole.

That said, yes, I can open a tracking issue getting the vmfirmwareigvm_dll crate cross-compiling.

Reply:
Where is the evidence for "is good for reading only"? Here is an example in the object crate source code how to copy PE files https://github.com/gimli-rs/object/blob/fb4446b56e6b4097afb057f4b5ea68ecc11227a5/crates/examples/src/bin/pecopy.rs which means writing is supported. I can't immediately see what "misc cruft" would be present in the resource DLL, might want to clarify in the issue. Thanks for grappling with building the dll!

Reply:
Copying != Building from scratch. Its easy to use `object` to read / parse an existing file (and all the "misc cruft" inside it, such as the DOS header, section headers, tables, etc...), and then dump its contents out to rebuilt it. Its a whole different ball-game to build and compose all those structs out of thin air to craft a new DLL file.

Not saying its impossible / overly-difficult, just saying that is a non-negligible amount of additional effort. This is what I meant by "it won't get you all the way there".

That approach may certainly be a viable way to get this functionality working in a cross-platform way, but its not the only way.

If you're inclined to get nerd-sniped here and whip up tool Roman, I'd love to see what you come up with 😄

Reply:
Tracked via #314

Reply:
Thanks 🤣 Can already build the `res` file:
```
x86_64-w64-mingw32-windres -J rc -i /home/krom/src/hvlite-ms/oss/openhcl/vmfirmwareigvm_dll/resources.rc -O res -o vmfw.res 
```

Now I might try stuffing it into the resource section of a stub DLL